csharp/extractor/Semmle.Util/FileUtils.cs

```diff
@@ -144,20 +144,40 @@ public static string NestPaths(ILogger logger, string? outerpath, string innerpa
             return nested;
         }
 
+        private static string? tempFolderPath = null;
+        private static readonly object lockObject = new();
+
         public static string GetTemporaryWorkingDirectory(Func<string, string?> getEnvironmentVariable, string lang, out bool shouldCleanUp)
         {
             shouldCleanUp = false;
             var tempFolder = getEnvironmentVariable($"CODEQL_EXTRACTOR_{lang}_SCRATCH_DIR");
+            if (!string.IsNullOrEmpty(tempFolder))
+            {
+                return tempFolder;
+            }
 
-            if (string.IsNullOrEmpty(tempFolder))
+            if (!string.IsNullOrEmpty(tempFolderPath))
             {
+                shouldCleanUp = true;
+                return tempFolderPath;
+            }
```

Comment:
I think this can be removed (I don't think it will cause any real performance issues only having this block under the "lock".

Reply:
Could you also briefly explain where there is a race condition that requires a lock?

Reply:
I think in practice, we would never see a case where the lock is required because of the order how we construct the temporary directories. But `DownloadMissingPackages` does have a `Parallel.ForEach` with concurrent access to the temp directory location, so if we removed/reordered a couple of things, it might be required.